File: Sources/TokenTrackerCore/UsageAlertEvaluator.swift
import Foundation

public struct UsageAlertSettings: Equatable, Sendable {
    public let notificationsEnabled: Bool
    public let fiveHourThreshold: Int
    public let sevenDayThreshold: Int
    public let resetWarningMinutes: Int

    public init(notificationsEnabled: Bool, fiveHourThreshold: Int, sevenDayThreshold: Int, resetWarningMinutes: Int) {
        self.notificationsEnabled = notificationsEnabled
        self.fiveHourThreshold = max(0, min(100, fiveHourThreshold))
        self.sevenDayThreshold = max(0, min(100, sevenDayThreshold))
        self.resetWarningMinutes = max(0, resetWarningMinutes)
    }
}

public struct UsageAlertCandidate: Equatable, Sendable {
    public let id: String
    public let title: String
    public let body: String

    public init(id: String, title: String, body: String) {
        self.id = id
        self.title = title
        self.body = body
    }
}

public enum UsageAlertEvaluator {
    public static func candidates(
        snapshot: UsageSnapshot,
        settings: UsageAlertSettings,
        now: Date = Date(),
        localizer: Localizer = Localizer(language: .english)
    ) -> [UsageAlertCandidate] {
        guard settings.notificationsEnabled else {
            return []
        }

        return Provider.allCases.flatMap { provider in
            let usage = snapshot.usage(for: provider)
            return candidates(
                for: usage,
                settings: settings,
                now: now,
                localizer: localizer
            )
        }
    }

    private static func candidates(
        for usage: ProviderUsage,
        settings: UsageAlertSettings,
        now: Date,
        localizer: Localizer
    ) -> [UsageAlertCandidate] {
        guard usage.isAvailable else {
            return []
        }

        var alerts: [UsageAlertCandidate] = []
        if let remaining = usage.remainingPercent5h,
           settings.fiveHourThreshold > 0,
           remaining <= settings.fiveHourThreshold {
            alerts.append(
                UsageAlertCandidate(
                    id: "\(usage.provider.rawValue)-5h-low",
                    title: localizer.text(.fiveHourAlertTitle),
                    body: "\(usage.provider.displayName) 5h \(remaining)% <= \(settings.fiveHourThreshold)%"
                )
            )
        }

        if let remaining = usage.remainingPercent7d,
           settings.sevenDayThreshold > 0,
           remaining <= settings.sevenDayThreshold {
            alerts.append(
                UsageAlertCandidate(
                    id: "\(usage.provider.rawValue)-7d-low",
                    title: localizer.text(.sevenDayAlertTitle),
                    body: "\(usage.provider.displayName) 7d \(remaining)% <= \(settings.sevenDayThreshold)%"
                )
            )
        }

        alerts.append(contentsOf: resetAlerts(for: usage, settings: settings, now: now, localizer: localizer))
        return alerts
    }

    private static func resetAlerts(
        for usage: ProviderUsage,
        settings: UsageAlertSettings,
        now: Date,
        localizer: Localizer
    ) -> [UsageAlertCandidate] {
        guard settings.resetWarningMinutes > 0 else {
            return []
        }

        let warningWindow = TimeInterval(settings.resetWarningMinutes * 60)
        let resets: [(label: String, date: Date?)] = [
            ("5h", usage.resetAt5h),
            ("7d", usage.resetAt7d)
        ]

        return resets.compactMap { reset in
            guard let date = reset.date else {
                return nil
            }
            let remaining = date.timeIntervalSince(now)
            guard remaining > 0, remaining <= warningWindow else {
                return nil
            }
            let minutes = max(0, Int(ceil(remaining / 60)))
            let resetID = Int(date.timeIntervalSince1970)
            return UsageAlertCandidate(
                id: "\(usage.provider.rawValue)-\(reset.label)-reset-\(resetID)",
                title: localizer.text(.resetAlertTitle),
                body: "\(usage.provider.displayName) \(reset.label) reset in \(minutes)m"
            )
        }
    }
}

extension UsageSnapshot {
    public func usage(for provider: Provider) -> ProviderUsage {
        switch provider {
        case .claude:
            return claude
        case .codex:
            return codex
        }
    }
}

File: windows/TokenTracker.Windows.Core/UsageAlertEvaluator.cs
namespace TokenTracker.Windows.Core;

public sealed record UsageAlertSettings
{
    public UsageAlertSettings(
        bool notificationsEnabled,
        int fiveHourThreshold,
        int sevenDayThreshold,
        int resetWarningMinutes)
    {
        NotificationsEnabled = notificationsEnabled;
        FiveHourThreshold = Math.Clamp(fiveHourThreshold, 0, 100);
        SevenDayThreshold = Math.Clamp(sevenDayThreshold, 0, 100);
        ResetWarningMinutes = Math.Max(0, resetWarningMinutes);
    }

    public bool NotificationsEnabled { get; }
    public int FiveHourThreshold { get; }
    public int SevenDayThreshold { get; }
    public int ResetWarningMinutes { get; }
}

public sealed record UsageAlertCandidate(string Id, string Title, string Body);

public static class UsageAlertEvaluator
{
    public static IReadOnlyList<UsageAlertCandidate> Candidates(
        UsageSnapshot snapshot,
        UsageAlertSettings settings,
        DateTimeOffset? now = null,
        Localizer? localizer = null)
    {
        if (!settings.NotificationsEnabled)
        {
            return Array.Empty<UsageAlertCandidate>();
        }

        localizer ??= new Localizer(AppLanguage.English);
        var current = now ?? DateTimeOffset.Now;
        return new[]
            {
                snapshot.Usage(Provider.Claude),
                snapshot.Usage(Provider.Codex)
            }
            .SelectMany(usage => Candidates(usage, settings, current, localizer))
            .ToArray();
    }

    private static IEnumerable<UsageAlertCandidate> Candidates(
        ProviderUsage usage,
        UsageAlertSettings settings,
        DateTimeOffset now,
        Localizer localizer)
    {
        if (!usage.IsAvailable)
        {
            yield break;
        }

        if (usage.RemainingPercent5h is { } fiveHourRemaining &&
            settings.FiveHourThreshold > 0 &&
            fiveHourRemaining <= settings.FiveHourThreshold)
        {
            yield return new UsageAlertCandidate(
                $"{usage.Provider.ToId()}-5h-low",
                localizer.Text(L10nKey.FiveHourAlertTitle),
                $"{usage.DisplayName} 5h {fiveHourRemaining}% <= {settings.FiveHourThreshold}%");
        }

        if (usage.RemainingPercent7d is { } sevenDayRemaining &&
            settings.SevenDayThreshold > 0 &&
            sevenDayRemaining <= settings.SevenDayThreshold)
        {
            yield return new UsageAlertCandidate(
                $"{usage.Provider.ToId()}-7d-low",
                localizer.Text(L10nKey.SevenDayAlertTitle),
                $"{usage.DisplayName} 7d {sevenDayRemaining}% <= {settings.SevenDayThreshold}%");
        }

        foreach (var candidate in ResetAlerts(usage, settings, now, localizer))
        {
            yield return candidate;
        }
    }

    private static IEnumerable<UsageAlertCandidate> ResetAlerts(
        ProviderUsage usage,
        UsageAlertSettings settings,
        DateTimeOffset now,
        Localizer localizer)
    {
        if (settings.ResetWarningMinutes <= 0)
        {
            yield break;
        }

        var warningWindow = TimeSpan.FromMinutes(settings.ResetWarningMinutes);
        foreach (var reset in new[] { ("5h", usage.ResetAt5h), ("7d", usage.ResetAt7d) })
        {
            if (reset.Item2 is not { } date)
            {
                continue;
            }

            var remaining = date - now;
            if (remaining <= TimeSpan.Zero || remaining > warningWindow)
            {
                continue;
            }

            var minutes = Math.Max(0, (int)Math.Ceiling(remaining.TotalMinutes));
            var resetId = date.ToUnixTimeSeconds();
            yield return new UsageAlertCandidate(
                $"{usage.Provider.ToId()}-{reset.Item1}-reset-{resetId}",
                localizer.Text(L10nKey.ResetAlertTitle),
                $"{usage.DisplayName} {reset.Item1} reset in {minutes}m");
        }
    }
}

public static class UsageSnapshotExtensions
{
    public static ProviderUsage Usage(this UsageSnapshot snapshot, Provider provider) => provider switch
    {
        Provider.Claude => snapshot.Claude,
        Provider.Codex => snapshot.Codex,
        _ => throw new ArgumentOutOfRangeException(nameof(provider), provider, null)
    };

    public static string ToId(this Provider provider) => provider switch
    {
        Provider.Claude => "claude",
        Provider.Codex => "codex",
        _ => provider.ToString().ToLowerInvariant()
    };
}
